### PR TITLE
Go SDK: Update memfs to parse the List() pattern as a glob, not a regexp

### PR DESCRIPTION
**Pull request by gonzojive**
This makes the memfs implementation consistent with the filesystem
implementation of List().

The docstring for filesystem.Interface.List does not specify how the pattern
should be interpretted. It says: "List expands a pattern to a list of
filenames." Perhaps that docstring should be updated to be more specific.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Review by damccorm — COMMENTED** [review]
I'll ask @lostluck to weigh in here since I'm not super up to date on how aggressively we want to work to unify our IO behavior, but I don't think we can take this since it will break any existing users (especially since its breaking in a silent/potential data-lossy way).

In a vacuum I agree that unifying behavior is a good idea, so this could be a good candidate if we upgrade major versions eventually, or if we can find a less breaking way to do this (which seems kinda unlikely)

**Comment by lostluck** [comment]
In this instance, I think we should change it.

1. We don't really demonstrate usage of the memfs anywhere. https://github.com/apache/beam/search?l=Go&q=memfs
2. We don't have properly defined behavior.
3. Regex in this case is surprising, especially if coming from the local or gcs file systems.

Looking at the local and gcs implementations, they use filepath.Match/Glob instead of plain path.Match. filepath is OS separator aware.

A github search doesn't reveal any usage, and neither does the [pkg page for v2](https://pkg.go.dev/github.com/apache/beam/sdks/v2/go/pkg/beam/io/filesystem/memfs?tab=importedby).  Checking the pre-modules version has [1 package ](https://pkg.go.dev/github.com/apache/beam/sdks/go/pkg/beam/io/filesystem/memfs?tab=importedby) that hasn't updated in 3 years, and doesn't use it explicitly (but does use local and gcs).

As a result, I think a move towards consistency, and documenting the behavior is the right one.

File: sdks/go/pkg/beam/io/filesystem/memfs/memory.go [hunk 1]
```diff
@@ -54,14 +55,18 @@ func (f *fs) List(_ context.Context, glob string) ([]string, error) {
...
-		return nil, err
+	globNoScheme := strings.TrimPrefix(glob, "memfs://")
+	if globNoScheme == glob {
+		return nil, fmt.Errorf("invalid pattern %q does not have prefix memfs://", glob)
 	}
 
 	var ret []string
 	for k := range f.m {
-		if pattern.MatchString(k) {
+		matched, err := path.Match(globNoScheme, strings.TrimPrefix(k, "memfs://"))
```
Comment (lostluck) [inline sdks/go/pkg/beam/io/filesystem/memfs/memory.go:65]: Please switch this to (path/filepath) filepath.Match instead, for consistency with the gcs and local implementations. In practice it's always linux.  Please also update the documentation in filesystem.Interface.List to be explicit as you suggested.

Thank you for the contribution!

**Review by lostluck — CHANGES_REQUESTED** [review]
Small change request, but otherwise this LGTM.

File: sdks/go/pkg/beam/io/filesystem/memfs/memory.go (hunk 1 above)
Comment (gonzojive) [reply sdks/go/pkg/beam/io/filesystem/memfs/memory.go]: Done.

Please note that I also made the pattern accepted by List() more tolerant. It no longer requires a memfs:// prefix.

**Comment by gonzojive** [comment]
I'm not sure what the etiquette is for resolving comments on github, but I think I addressed the comment, so I resolved it.

**Review by lostluck — APPROVED** [review]
LGTM
Thank your for your patience and sorry for the delay! This one slipped through the cracks.

Beam's usual standard is if the comment is to mark the comment resolved, which at least also emails folks subscribed to the issue.

**Pull request merged**